### ISSUE TITLE
Oracle.ManagedDataAccess.Client errors when performing DatabaseReader.ReadAllSchema

### ISSUE THREAD

**Issue by rufio620** [martinjw/dbschemareader#13]
It appears that this the library has issues when reading schema using the oracle managed driver.

https://www.nuget.org/packages/Oracle.ManagedDataAccess/

The following code snippet results in an exception.

```c#
var connString = @"Data Source=(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=192.168.137.10)(PORT=1521)))(CONNECT_DATA=(SERVER=DEDICATED)(SERVICE_NAME=XE)));User Id=MySchemaUser;Password=MyPass;";
var providerName = "Oracle.ManagedDataAccess.Client";

using (var schemaReader = new DatabaseReader(connString, providerName))
{
	schemaReader.ReadAll();
}
```

**Message**
ORA-01008: not all variables bound
**Stack Trace**
 at OracleInternal.ServiceObjects.OracleCommandImpl.VerifyExecution(OracleConnectionImpl connectionImpl, Int32& cursorId, Boolean bThrowArrayBindRelatedErrors, OracleException& exceptionForArrayBindDML, Boolean& hasMoreRowsInDB, Boolean bFirstIterationDone)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ExecuteReader(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, OracleDataReaderImpl& rdrImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[] scnForExecution, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, Int64& internalInitialLOBFS, OracleException& exceptionForArrayBindDML, Boolean isDescribeOnly, Boolean isFromEF)
   at Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteReader(Boolean requery, Boolean fillRequest, CommandBehavior behavior)
   at Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at System.Data.Common.DbCommand.ExecuteReader()
   at DatabaseSchemaReader.ProviderSchemaReaders.Databases.SqlExecuter.ExecuteDbReader(DbConnection connection)
   at DatabaseSchemaReader.ProviderSchemaReaders.Adapters.OracleAdapter.Tables(String tableName)
   at DatabaseSchemaReader.ProviderSchemaReaders.Builders.TableBuilder.Execute(CancellationToken ct)
   at DatabaseSchemaReader.DatabaseReader.AllTables(CancellationToken ct)
   at DatabaseSchemaReader.DatabaseReader.ReadAll(CancellationToken ct)
   at DatabaseSchemaReader.DatabaseReader.ReadAll()
   at UserQuery
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()

**Comment by rufio620** [comment]
This also appears to be happening with the unmanaged driver as well.

I took a look at the source code and I think I found the root cause of this.

There are two places in the code where reader attempts to set the BindByName on the oracle data provider via reflection.

One in OracleSchemaReader.cs and in OracleSqlExecuter.cs both static methods with EnsureOracleBindByName

The one in OracleSqlExecuter.cs has the following code

```c#
protected void EnsureOracleBindByName(DbCommand command)
        {
            //Oracle.DataAccess.Client only binds first parameter match unless BindByName=true
            //so we violate LiskovSP (in reflection to avoid dependency on ODP)
#if !COREFX
            var bindByName =  typeof(DbCommand).GetProperty("BindByName");
#else
            var bindByName = typeof(DbCommand).GetTypeInfo().GetDeclaredProperty("BindByName");
#endif           
            if (bindByName != null)
            {
                bindByName.SetValue(command, true, null);
            }
        }
```

This looks like it should be changed to command.GetProperty("BindByName")
The current code will always return null for the BindByName property because it is looking at the base type not the actual provider command type. The version in OracleSchemaReader.cs appears to be implemented correctly as it references the passed in command object to get the type information.

**Comment by martinjw** [comment]
Yes, that's right - instead of typeof(DbCommand) it should be command.GetType()
Fix coming soon.

**Comment by rufio620** [comment]
Awesome! Any idea when the next nuget release will be?

**Comment by martinjw** [comment]
I'll probably do a release at the weekend, if not earlier.

**Comment by martinjw** [comment]
@rufio620  Released as v2.1.0
